mongowire/sock.go

```diff
@@ -1,22 +1,43 @@
 package mongowire
 
 import (
+	"context"
 	"io"
 
 	"github.com/pkg/errors"
 )
 
 const MaxInt32 = 2147483647
 
-func ReadMessage(reader io.Reader) (Message, error) {
+func ReadMessage(ctx context.Context, reader io.Reader) (Message, error) {
 	// read header
 	sizeBuf := make([]byte, 4)
-	n, err := reader.Read(sizeBuf)
-	if err != nil {
-		return nil, errors.WithStack(err)
+	type readResult struct {
+		n   int
+		err error
 	}
-	if n != 4 {
-		return nil, errors.Errorf("didn't read message size from socket, got %d", n)
+
+	doRead := func(ctx context.Context, readFinished chan readResult, out []byte) {
+		defer close(readFinished)
+		n, err := reader.Read(out)
+		select {
+		case readFinished <- readResult{n: n, err: err}:
+		case <-ctx.Done():
+		}
+	}
+
+	readFinished := make(chan readResult)
+	go doRead(ctx, readFinished, sizeBuf)
```

Comment:
I'd just call this inline rather than calling it as a function.

Reply:
Copy-pasted `doRead` into the two places it's used.